fitbenchmarking/parsing/cutest_parser.py

```diff
@@ -0,0 +1,240 @@
+"""
+This file calls the pycutest interface for SIF data
+"""
+
+from __future__ import print_function
+
+import numpy as np
+import os
+import pycutest
+try:
+    from tempfile import TemporaryDirectory
+except ImportError:
+    from backports.tempfile import TemporaryDirectory
+
+from collections import OrderedDict
+from fitbenchmarking.parsing.base_parser import Parser
+from fitbenchmarking.parsing.fitting_problem import FittingProblem
+from fitbenchmarking.utils.create_dirs import del_contents_of_dir
+from fitbenchmarking.utils.logging_setup import logger
+
+
+# Empty the cache
+del_contents_of_dir(os.environ["PYCUTEST_CACHE"])
+
+
+class CutestParser(Parser):
+    """
+    Use the pycutest interface to parse SIF files
+    """
+
+    def parse(self):
+        """
+        Get data into a Fitting Problem via cutest.
+
+        :return: The fully parsed fitting problem
+        :rtype: fitbenchmarking.parsing.fitting_problem.FittingProblem
+        """
+        self.mastsif_dir = TemporaryDirectory()
+
+        # set the MASTSIF environment variable so that pycutest
+        # can find the sif files
+        os.environ["MASTSIF"] = self.mastsif_dir.name
+
+        self._n = None
+        self._m = None
+
+        # get just the short filename (minus the .SIF)
+        fp = FittingProblem()
+
+        # Collect x and create new file with blank y
+        fname, fp.data_x, fp.data_y, fp.data_e = self._setup_data()
+
+        self._p = self._import_problem(fname)
+
+        fp.name = self._p.name
+        self._y = fp.data_y
+
+        fp.function = self._function  # self._p.obj
+        fp.equation = None
+        fp.starting_values = self._get_starting_values()
+        fp.start_x = None
+        fp.end_x = None
+
+        # Use the object id as a hash to the function.
+        # In order to prevent the id being reused also store the x_data
+        self._cache = {id(fp.data_x): {'f': self._p.objcons, 'x': fp.data_x}}
```

Comment:
The issue we had is that x data could be very large and doing a comparison would have a time overhead (not good for benchmarking). So I was trying to use a faster comparison.

Ideally this would be very fast to compute so using the object id is pretty good.

We also know that the data object doesn't change throughout the main benchmarking (it does change when plotting/testing but time is not an issue then), so I thought this would be suitable.

Reply:
I guess the point I'm confused by is when would you be doing this comparison?  What is the series of events that needs to occur for `_function` to have already been called on `x`?

Reply:
It happens whenever the software runs a new set of parameters. We fit `| y - f(x, p) |` where `x` and `y` are fixed, but for the purpose of plotting/testing, it's necessary to be able change `x`. During the fitting, `x` does not change and is passed in as `fitting_problem.data_x` which will always be the same object, however `p` varies until a local minima is reached.

Reply:
I'm more than happy to use a different way of achieving a fast comparison, however I can't think of a better alternative.

Reply:
Ok so basically within a single problem definition it will never enter the `Exception` because x is always constant?  But when changing to a new problem definition it will always change because it will use a different x (or the same x but with a different location in memory)?  Is that correct?

Reply:
Not quite, different problem definitions use separate parsers (and thereby they have their own caches).

During benchmarking, every minimizer uses the same fitting problem (and the x from it), so during benchmarking for a single problem it will not change. It will change when plotting as we sort x before evaluating to ensure the line plot is sensible. In future, we might over-sample x in plotting as well in order to get smooth curves.

As an example of the cache when fitting problems `t1.SIF` and `t2.SIF` with minimizers a, b, and c would give:
```
parse problem t1.SIF (add fitting_problem1.x to t1 cache)
fit t1.SIF using a (cache hit on fitting_problem1.x) <- function could be called many times as part of the fitting, but always with the same object
fit t1.SIF using b (cache hit on fitting_problem1.x)
fit t1.SIF using c (cache hit on fitting_problem1.x)
parse problem t2.SIF (add fitting_problem2.x to t2 cache)
fit t2.SIF using a (cache hit on fitting_problem2.x)
fit t2.SIF using b (cache hit on fitting_problem2.x)
fit t2.SIF using c (cache hit on fitting_problem2.x)
plot result for t1.SIF using a (t1 cache miss, add sorted(fitting_problem1.x)) <- sorted(fitting_problem1.x) stored in plotting object so the same object will be used each time
plot result for t1.SIF using b (cache hit on sorted(fitting_problem1.x))
plot result for t1.SIF using c (cache hit on sorted(fitting_problem1.x))
plot result for t2.SIF using a (t2 cache miss, add sorted(fitting_problem2.x))
plot result for t2.SIF using b (cache hit on sorted(fitting_problem2.x))
plot result for t2.SIF using c (cache hit on sorted(fitting_problem2.x))
```